common/src/main/proto/signup.proto

```diff
@@ -0,0 +1,60 @@
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "com.rtbhouse.grpc.loadbalancer";
+option java_outer_classname = "SignupProto";
+
+package com.rtbhouse.grpc.loadbalancer;
+
+service ServerSignup {
+    // Bidirectional rpc for registering server in load balancer.
+
+    // A server calls the rpc when it wants to be included in load balancing policy.
+    // The following life-cycle is involved:
+    // * In the first message, the server should include 'server_details' field
+    //   and 'ready_to_serve' = true.
+    // * The first message received by the server, contains 'heartbeats_frequency'
+    //   field. From then on, in order to stay on the LB list, the server sends
+    //   heartbeats (messages containing only 'ready_to_serve' field) at regular
+    //   intervals indicated by 'heartbeats_frequency'. Note that server appears
+    //   on the list after sending the first "real" heartbeat (second message).
+    // * When the server becomes unhealthy, or simply ends its work, and wants
+    //   to be removed from the LB policy immediately, it should call onCompleted().
+    //   Besides, when LB doesn't get a heartbeat after 'heartbeats_frequency' time,
+    //   it also deletes the server from the list, but LB doesn't call onCompleted():
+    //   if delayed heartbeat will be delivered, server will reappear on the LB list.
+    rpc Signup (stream ServerReport) returns (stream LoadBalancerSignupReply) {
+    }
+}
+
+// Every message should include 'ready_to_serve' field.
+// The field 'server_details' is obligatory only for the first message
+// being sent in the rpc.
+message ServerReport {
+    bool ready_to_serve = 1;
```

Comment:
Obecność tej flagi sugerowałaby, że serwer może nie być "ready", jednak w SingleLoadBalanceConnectorze nigdzie nie widzę, żeby taka wartość tej flagi była ustawiana. Z tego co widzę to przy signupie od razu ustawiacie to na true, a każdy kolejny heartbeat również to ustawia. Może ta flaga nie ma sensu w takim przypadku i wystarczy tu samo server_details?

Reply:
To prawda, trzeba byłoby się zdecydować:
- albo na usunięcie tej flagi (wtedy heartbeaty byłyby pustymi wiadomościami, bo z nimi nie leci server_details, trzeba byłoby się upewnić że w gRPC to nie jest problem),
- albo na prawidłową jej walidację po stronie LB (teraz jest ona ignorowana co jest bez sensu), i w momencie gdy przyszedłby heartbeat z wartością false, to rzeczywiście usuwać serwer z listy.